tensorflow_quantum/core/ops/tfq_adj_grad_op.cc

```diff
@@ -0,0 +1,403 @@
+    using State = StateSpace::State;
+
+    auto DoWork = [&](int start, int end) {
+      // Begin simulation.
+      int largest_nq = 1;
+      State sv = StateSpace(largest_nq, tfq_for).CreateState();
+      State scratch = StateSpace(largest_nq, tfq_for).CreateState();
+      State scratch2 = StateSpace(largest_nq, tfq_for).CreateState();
+
+      for (int i = start; i < end; i++) {
+        int nq = num_qubits[i];
+        Simulator sim = Simulator(nq, tfq_for);
+        StateSpace ss = StateSpace(nq, tfq_for);
+        if (nq > largest_nq) {
+          // need to switch to larger statespace.
+          largest_nq = nq;
+          sv = ss.CreateState();
+          scratch = ss.CreateState();
+          scratch2 = ss.CreateState();
+        }
+
+        // (#679) Just ignore empty program
+        if (qsim_circuits[i].gates.size() == 0) {
+          continue;
+        }
+
+        ss.SetStateZero(sv);
+        for (int j = 0; j < partial_fused_circuits[i].size(); j++) {
+          for (int k = 0; k < partial_fused_circuits[i][j].size(); k++) {
+            qsim::ApplyFusedGate(sim, partial_fused_circuits[i][j][k], sv);
+          }
+          if (j == partial_fused_circuits[i].size() - 1) {
+            break;
+          }
+          qsim::ApplyGate(
+              sim, qsim_circuits[i].gates[gradient_gates[i][j].index], sv);
+        }
+
+        // sv now contains psi
+        // scratch contains (sum_i paulis_sums[i] * downstream_grads[i])|psi>
+        // scratch2 now contains psi as well.
+        AccumulateOperators(pauli_sums[i], downstream_grads[i], sim, ss, sv,
+                            scratch2, scratch);
+
+        for (int j = partial_fused_circuits[i].size() - 1; j >= 0; j--) {
+          for (int k = partial_fused_circuits[i][j].size() - 1; k >= 0; k--) {
+            ApplyFusedGateDagger(sim, partial_fused_circuits[i][j][k], sv);
+            ApplyFusedGateDagger(sim, partial_fused_circuits[i][j][k], scratch);
+          }
+          if (j == 0) {
+            // last layer will have no parametrized gates so can break.
+            break;
+          }
+
+          // Hit a parameterized gate.
+          ApplyGateDagger(
+              sim, qsim_circuits[i].gates[gradient_gates[i][j - 1].index], sv);
+          for (int k = 0; k < gradient_gates[i][j - 1].grad_gates.size(); k++) {
+            // Copy sv onto scratch2 in anticipation of non-unitary "gradient
+            // gate".
+            ss.CopyState(sv, scratch2);
+            qsim::ApplyGate(sim, gradient_gates[i][j - 1].grad_gates[k], sv);
+
+            // don't need not-found check since this is done upstream already.
+            const auto it = maps[i].find(gradient_gates[i][j - 1].params[k]);
+            const int loc = it->second.first;
+            (*output_tensor)(i, loc) += ss.RealInnerProduct(sv, scratch) +
+                                        ss.RealInnerProduct(scratch, sv);
+
+            ss.CopyState(scratch2, sv);
```

Comment:
it seems like applying the `gradient_gates[i][j-1].grad_gates[k]` to the statevector, and calculate the innerproduct and then rollback the statevector before applying the `grad_gates[k]`. It looks like `scratch2` is always a temporal backup of `sv`. If it is the case, shall we reduce the number of `CopyState()` from `2` to `1`? I mean:
```cc
ss.CopyState(sv, scratch2);
qsim::ApplyGate(sim, gradient_gates[i][j - 1].grad_gates[k], scratch2); /* it was `sv`*/

// don't need not-found check since this is done upstream already.
const auto it = maps[i].find(gradient_gates[i][j - 1].params[k]);
const int loc = it->second.first;
(*output_tensor)(i, loc) += ss.RealInnerProduct(scratch2, scratch) +
                            ss.RealInnerProduct(scratch, scratch2);   /* <scratch2 | scratch> + <scratch|scratch2> */
```

Reply:
I don't think we can get rid of the `CopyState`. The code inside this for loop runs over the number of gradient_gate objects we have for a single gate. For a gate like FSIM which has 2 parameters, we must apply the gradient gate for the first parameter, compute the inner product, undo, apply the gradient gate for the second parameter, undo and then we can continue. This `for (int k = 0; k < gradient_gates[i][j - 1].grad_gates.size(); k++)` loop might run twice, and when it does that is when we need the copy.... _I think_ . Does this make sense ?

Reply:
Yes it does make sense, but I am just talking about the loop variance... Sorry, could you please help me to understand the code more?
Here is the original code with my comments on the loop variance.
```cc
        for (int j = partial_fused_circuits[i].size() - 1; j >= 0; j--) {
          for (int k = partial_fused_circuits[i][j].size() - 1; k >= 0; k--) { /* initialize sv, scratch */ }
          if (j == 0) break;

          ApplyGateDagger(
              sim, qsim_circuits[i].gates[gradient_gates[i][j - 1].index], sv); 
          /* Let `sv_before` == `sv`, `scratch_before` == `scratch` before this for-loop with k */
          for (int k = 0; k < gradient_gates[i][j - 1].grad_gates.size(); k++) {
            ss.CopyState(sv, scratch2); // sv == scratch2 == `sv_before` always due to loop invariance.
            qsim::ApplyGate(sim, gradient_gates[i][j - 1].grad_gates[k], sv); // this sv is ahead of `sv_before` with gradient_gates[i][j].grad_gates[k]. let `sv_after`
            // don't need not-found check since this is done upstream already.
            const auto it = maps[i].find(gradient_gates[i][j - 1].params[k]);
            const int loc = it->second.first;
            (*output_tensor)(i, loc) += ss.RealInnerProduct(sv, scratch) +
                                        ss.RealInnerProduct(scratch, sv); // scratch == scratch_before always. sv == sv_after
            ss.CopyState(scratch2, sv); // this undo sv's latest ApplyGate() above inside this for-loop. now sv == sv_before.
          }
          // Loop invariance : for-loop always finishes at every step with
          // sv == `sv_before`
          // scratch == `scratch_before`
          // scratch2 == `sv_before`
          ApplyGateDagger(
              sim, qsim_circuits[i].gates[gradient_gates[i][j - 1].index],
              scratch);
        }
      }
```

And here is my suggestion. To me, regardless of what gate is applied and how many of it is applied, for-loop invariance confirms that the values of `sv` and `scratch` are preserved (they are both `sv_before` and `scratch_before`), and the internal inner product values are well guaranteed with the same arguments (scratch & sv_after I mentioned). I think it also does the same undoing effect on the source `sv` for the FSIM gate with 2 parameters, because there is no change at the loop variables... 
```cc
        for (int j = partial_fused_circuits[i].size() - 1; j >= 0; j--) {
          for (int k = partial_fused_circuits[i][j].size() - 1; k >= 0; k--) { /* initialize sv, scratch */ }
          if (j == 0) break;

          ApplyGateDagger(
              sim, qsim_circuits[i].gates[gradient_gates[i][j - 1].index], sv); 
          /* say `sv_before` == `sv` `scratch_before` == `scratch` before this for-loop */
          for (int k = 0; k < gradient_gates[i][j - 1].grad_gates.size(); k++) {
            ss.CopyState(sv, scratch2); // sv == scratch2 == `sv_before` always due to loop invariance.
            qsim::ApplyGate(sim, gradient_gates[i][j - 1].grad_gates[k], scratch2); // this scratch2 is ahead of `sv_before` with gradient_gates[i][j].grad_gates[k]. let `sv_after`
            // don't need not-found check since this is done upstream already.
            const auto it = maps[i].find(gradient_gates[i][j - 1].params[k]); // regardless of what sv, scratch, scratch2 are.
            const int loc = it->second.first;
            (*output_tensor)(i, loc) += ss.RealInnerProduct(scratch2, scratch) +
                                        ss.RealInnerProduct(scratch, scratch2); // scratch == scratch_before always. scratch2 == sv_after
             // sv == sv_before without additional ss.CopyState.
          }
          // Loop invariance : for-loop always finishes at every step with
          // sv == `sv_before`
          // scratch == `scratch_before`
          // scratch2 == `sv_after` at the latest gradient_gate, but not used anywhere because it is overwritten by the `sv_before` of the next step due to `ss.CopyState(sv, scratch2)`
          ApplyGateDagger(
              sim, qsim_circuits[i].gates[gradient_gates[i][j - 1].index],
              scratch);
        }
      }
```

Reply:
Thanks for discussion online!